File: sdk/python/examples/basic_bot.py
from agentnews import AgentNewsClient
import nacl.signing
import base58
import time

# --- SETUP FOR EXAMPLE ---
# In production, load your secure private key.
# For demo, generate a fresh keypair and display it.
signing_key = nacl.signing.SigningKey.generate()
# Solana private keys are 64 bytes (32 byte secret + 32 byte public)
secret_key_bytes = signing_key.encode() + signing_key.verify_key.encode()
agent_private_key_str = base58.b58encode(secret_key_bytes).decode('utf-8')
agent_public_key_str = base58.b58encode(signing_key.verify_key.encode()).decode('utf-8')

print("🤖 Starting Autonomous Python Trading Agent...")
print(f"🔑 Agent Public Key: {agent_public_key_str}")
print("-" * 50)

# --- SDK INITIALIZATION ---
client = AgentNewsClient(
    private_key_base58=agent_private_key_str,
    api_url='http://localhost:3001' # Change to https://api.agentnewsapi.com for production
)

# --- EVENT HANDLERS ---
def on_connect():
    print("✅ Connected to Agent News Global Signal Firehose")
    try:
        balance = client.get_credit_balance()
        print(f"💰 Current Credit Balance: {balance} SOL")
        if balance == 0:
            print("\n⚠️ WARNING: Credit balance is 0.")
            print("Please send a tiny amount of SOL (e.g., 0.01) to:")
            print(f">> {agent_public_key_str} <<")
            print("The deposit will be automatically identified and credited.\n")
    except Exception as e:
        print(f"Failed to get balance: {e}")

def on_signal(signal):
    print("\n🚨 NEW GLOBAL SIGNAL INGESTED 🚨")
    print(f"Title: {signal.get('title')}")
    sig_score = signal.get('significance', {}).get('score')
    print(f"Significance Score: {sig_score}/100")
    sentiment = signal.get('sentiment', {})
    print(f"Sentiment: {sentiment.get('label')} ({sentiment.get('score')})")
    
    if sig_score and sig_score > 90:
        print(">> HIGH SIGNIFICANCE DETECTED. Executing macro trading strategy...")
    
    meta = signal.get('_meta')
    if meta:
        print(f"\n[Billing] This signal cost {meta.get('cost')} SOL.")
        print(f"[Billing] Remaining Balance: {meta.get('remainingCredits')} SOL")

def on_balance(new_balance):
    print(f"\n💳 Balance Updated: {new_balance} SOL remaining.")

def on_error(err):
    print(f"Error occurred: {err}")

# Wiring up events
client.on('connected', on_connect)
client.on('signal', on_signal)
client.on('balance', on_balance)
client.on('error', on_error)

# --- START THE AGENT ---
print("Connecting to Agent News API...")
try:
    client.connect()
    # Wait continuously
    while True:
         time.sleep(1)
except Exception as e:
    print(f"Failed to start agent: {e}")
    print("Ensure the Agent News API server is running.")

File: sdk/node/examples/basic-bot.js
const { AgentNewsClient } = require('../index');
const { Keypair } = require('@solana/web3.js');
const bs58 = require('bs58').default || require('bs58');

// --- SETUP FOR EXAMPLE ---
// In production, an agent would normally load its private key securely
// For this demo, we generate a fresh one and log it so you can fund it with devnet $SOL
const demoKeypair = Keypair.generate();
const agentPrivateKeyStr = bs58.encode(demoKeypair.secretKey);
const agentPublicKeyStr = demoKeypair.publicKey.toBase58();

console.log('🤖 Starting Autonomous Trading Agent...');
console.log('🔑 Agent Public Key:', agentPublicKeyStr);
console.log('--------------------------------------------------');

// --- SDK INITIALIZATION ---
// Initialize the client with the agent's private key
// The SDK automatically handles the zero-HITL auth to get an API Key
const client = new AgentNewsClient({
    privateKey: agentPrivateKeyStr,
    apiUrl: 'http://localhost:3001' // Change to https://api.agentnewsapi.com for production
});

// --- EVENT LISTENERS ---

// 1. Connection Event
client.on('connected', async () => {
    console.log('✅ Connected to Agent News Global Signal Firehose');
    
    try {
        const balance = await client.getCreditBalance();
        console.log(`💰 Current Credit Balance: ${balance} SOL`);
        
        if (balance === 0) {
            console.log(`\n⚠️  WARNING: Credit balance is 0.`);
            console.log(`Please send a tiny amount of SOL (e.g., 0.01) to:`);
            console.log(`>> ${agentPublicKeyStr} <<`);
            console.log(`The deposit will be automatically identified and credited.\n`);
        }
    } catch (err) {
        console.error('Failed to get balance:', err.message);
    }
});

// 2. The Core Value: Listening to Live Signals
client.on('signal', (signal) => {
    console.log('\n🚨 NEW GLOBAL SIGNAL INGESTED 🚨');
    console.log(`Title: ${signal.title}`);
    console.log(`Significance Score: ${signal.significance?.score}/100`);
    console.log(`Sentiment: ${signal.sentiment?.label} (${signal.sentiment?.score})`);
    
    // --- AUTONOMOUS LOGIC EXAMPLE ---
    if (signal.significance?.score > 90) {
        console.log('>> HIGH SIGNIFICANCE DETECTED. Executing macro trading strategy...');
        // executeTrade(signal);
    }
    
    // Log the cost of this signal processing
    if (signal._meta) {
        console.log(`\n[Billing] This signal cost ${signal._meta.cost} SOL.`);
        console.log(`[Billing] Remaining Balance: ${signal._meta.remainingCredits} SOL`);
    }
});

// 3. Balance Updates (Fired when deposits occur or credits are used)
client.on('balance', (newBalance) => {
    console.log(`\n💳 Balance Updated: ${newBalance} SOL remaining.`);
});

// 4. Error Handling
client.on('error', (err) => {
    if (err.code === 'INSUFFICIENT_CREDITS') {
        console.error('\n❌ STREAM PAUSED: Insufficient Credits.');
        console.error(`Please fund: ${agentPublicKeyStr}`);
    } else {
        console.error('WebSocket Error:', err.message);
    }
});

// --- START THE AGENT ---
console.log('Connecting to Agent News API...');
client.connect().catch(err => {
    console.error('Failed to start agent:', err.message);
    console.error('Ensure the Agent News API server is running locally on port 3001.');
});
